Traktor s3 2.3

ywwg opened the pull request:
Add support for the Traktor Kontrol S3

ywwg committed:
Traktor S3: Initial commit

Branched from Traktor S2 config
Most simple buttons and lights mapped.

ywwg committed:
Traktor S3: move from channel-specific to deck aliases

(does not work)

Be-ing committed:
HidController: loop until no more messages are available on poll

Otherwise, it seems messages build up in a queue which produces
a dramatic lag on Linux when moving faders quickly. This
regression was introduced between Mixxx 2.2 and 2.3 beta, likely
by switching HidController to nonblocking polling in PR #2179.

Be-ing committed:
HidController: do not copy incoming data before processing it

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: Fix most connections for deck switching

ywwg committed:
Traktor S3: Bind more of the library buttons

ywwg committed:
Traktor S3: Master Vus, start implementing color lights

ywwg committed:
Traktor S3: Lights behaving better, still a ways to go

Be-ing committed:
Merge remote-tracking branch 'upstream/2.3' into hid_polling

Be-ing committed:
HidController: filter redundant messages

ywwg committed:
Traktor S3: Try to get deck switching working

Be-ing committed:
ControllerManager: make kPollInterval a static member

so it can be referenced outside ControllerManager

Be-ing committed:
HidController: stop polling when exceeding poll interval

This solution avoids a deep copy of the incoming data.

Be-ing committed:
HidController: clang-format #includes

ywwg committed:
Traktor S3: good progress on lights

ywwg committed:
Traktor S3: more lights

ywwg committed:
Merge branch 'clone-deck-crash' into traktor-s3-2.3

ywwg committed:
Traktor S3: Fix color lights, add more

ywwg committed:
Traktor S3: configurable dim/bright color values

ywwg committed:
Traktor S3: pad buttons

ywwg committed:
Traktor S3: light up the reverse light

ywwg committed:
Traktor S3: Most functions working decently

ywwg committed:
Traktor S3: hotcue colors!

ywwg committed:
TraktorS3: fx buttons

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: clean up lint problems

ywwg committed:
Traktor S3: fix jog wheels

The Traktor jog wheels have their own time calculation, so we don't
need to use Mixxx's smoothing code.

ywwg committed:
TraktorS3: revert mistake diffs

ywwg committed:
Traktor S3: pitch slider relative mode

ywwg committed:
Traktor S3: animate segments around wheels

ywwg committed:
Traktor S3: Major refactor to be more objecty

ywwg committed:
Traktor S3: refactor continues, outputs starting to work again

ywwg committed:
Traktor S3: finish refactor!

ywwg committed:
Traktor S3: Fix some bugs. Pretty stable again

ywwg committed:
Traktor S3: better timers

ywwg committed:
Traktor S3: more rearranging (untested)

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: Make sure lights light up if triggered from gui, not just controller

ywwg committed:
Traktor S3: Fix hotcues

Also simplify some wrappers

ywwg committed:
Traktor S3: revert colormapper changes

ywwg committed:
Traktor S3: Make debugmode a var

Holzhaus commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,1694 @@
+///////////////////////////////////////////////////////////////////////////////////
+// JSHint configuration                                                          //
+///////////////////////////////////////////////////////////////////////////////////
+/* global controller                                                             */
+/* global HIDDebug                                                               */
+/* global HIDPacket                                                              */
+/* global HIDController                                                          */
+/* jshint -W016                                                                  */
```
```suggestion
```

We're using eslint now :D

ywwg replied:
when I remove this I get:

  20:27  error  'HIDController' is not defined  no-undef
   173:9   error  'HIDDebug' is not defined       no-undef
   383:9   error  'HIDDebug' is not defined       no-undef
  1059:9   error  'HIDDebug' is not defined       no-undef
  1098:28  error  'HIDPacket' is not defined      no-undef
  1099:27  error  'HIDPacket' is not defined      no-undef
  1319:23  error  'HIDPacket' is not defined      no-undef
  1320:23  error  'HIDPacket' is not defined      no-undef
  1595:9   error  'HIDDebug' is not defined       no-undef
  1604:17  error  'HIDDebug' is not defined       no-undef
  1609:17  error  'HIDDebug' is not defined       no-undef
  1614:13  error  'controller' is not defined     no-undef
  1648:9   error  'controller' is not defined     no-undef
  1651:5   error  'HIDDebug' is not defined       no-undef
  1669:5   error  'HIDDebug' is not defined       no-undef

ywwg replied:
ah, got it

Holzhaus commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,1694 @@
+        SKY: 0x28,
+        BLUE: 0x2C,
+        PURPLE: 0x30,
+        FUSCHIA: 0x34,
```
```suggestion
        FUCHSIA: 0x34,
```

Holzhaus commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,1694 @@
+            }
+            data[i][j] = b;
+        }
+        // if (i === 0) {
+        //     for (k = 0; k < data[0].length; k++) {
+        //         data[0][k] = 0x30;
+        //     }
+        // }
+        // for (d = 0; d < 8; d++) {
+        //     data[0][0x11+d] = (d+1) * 4 + 2;
+        // }
+        // for (d = 0; d < 8; d++) {
+        //     data[0][0x2A + d] = (d + 1) * 4 + 32 + 2;
+        // }
```
Remove?

ywwg replied:
done

ywwg committed:
Traktor S3: enable soft takeover for quickeffect

ywwg committed:
Update res/controllers/Traktor-Kontrol-S3-hid-scripts.js

remove obsolete hint config

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

ywwg committed:
Update res/controllers/Traktor-Kontrol-S3-hid-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

ywwg committed:
Traktor S3: notes

Holzhaus commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,1681 @@
+        0x0000CC: this.controller.LEDColors.BLUE,
+        0xCC00CC: this.controller.LEDColors.PURPLE,
+
+        0xCC0091: this.controller.LEDColors.FUSCHIA,
```
```suggestion
        0xCC0091: this.controller.LEDColors.FUCHSIA,
```

ywwg committed:
Update res/controllers/Traktor-Kontrol-S3-hid-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,1681 @@
+
+    this.debugMode = false;
+
+    // ==== Friendly User Configuration ====
```
Put this at the very top of the file before anything else. It can be a global variable.

ywwg replied:
done

ywwg committed:
Traktor S3: make user configs global

ywwg committed:
Traktor S3: make button brightnesses global too

ywwg commented:
Ah, I still need to figure out how to use the EXT button

ywwg committed:
Traktor S3: Implement channel 4 input mode

Fix some whoopses

ywwg commented:
cool I reverse engineered switching between line and mic!  It now works the same as I think it does in traktor

ywwg committed:
Traktor S3: remove redundant connections

ywwg committed:
Traktor S3: FX for input channel

ywwg committed:
Traktor S3: fix keylock light and input fx

microphone doesn't have a quickeffect.
Make master gain adjustable with shift+gain

ywwg committed:
Traktor S3: toggle whole chain

ywwg commented:
The controller just has enable/disable buttons per effect, and per deck.  They are combined with an AND operation -- every channel that has effects enabled will have all the effects that are selected.  If a channel has effects disabled, the knob will control the quickeffect.

Be-ing commented:
I still don't understand what you're saying or why you're trying to map the regular effects units to the QuickEffects.

ywwg commented:
each channel only has one FX knob and one FX button:

![image](https://user-images.githubusercontent.com/888940/90984424-dad1b480-e542-11ea-8702-05386353cd7b.png)

The fx section of the mixer only has on/off:

![image](https://user-images.githubusercontent.com/888940/90984430-e8873a00-e542-11ea-9088-004c10b2ac39.png)

It might also help to look at the s3 manual: 

![image](https://user-images.githubusercontent.com/888940/90984465-1cfaf600-e543-11ea-9edf-f6ced6c452f2.png)


![image](https://user-images.githubusercontent.com/888940/90984462-153b5180-e543-11ea-894f-0f61188eb63d.png)

ywwg commented:
(https://www.native-instruments.com/fileadmin/ni_media/downloads/manuals/traktor/TRAKTOR_KONTROL_S3_MK1_Manual_English_0120.pdf)

Be-ing commented:
I understand how it works in Traktor. Mixxx doesn't support that yet. That's what #2618 does. I don't understand the point of hacking some way to use those FX Select buttons before that feature exists in Mixxx.

ywwg commented:
This deck has very limited FX support.  It's essentially like a single FX unit with 4 slots + filter, and each channel is either on or off.  I decided to make it a little more flexible by having each channel either be quickeffect (fxenable off) or fx unit (fxenable on)

ywwg commented:
I don't understand how that PR changes things.  Or do you mean that the fx buttons should control the loading of saved chains?  I'm having the fx buttons enable/disable different effect units.  FX are a pretty major part of a dj controller, I don't like the idea of leaving them totally unmapped.  One alternative would be to have the knob only control the quickeffect, and not have any knob support for the FX units.

Be-ing commented:
> Or do you mean that the fx buttons should control the loading of saved chains?

Yes, that's what they're designed to do. They are meant to control the QuickEffects, not the regular effects units.

If you want to map a way to control effects chains, you could use the 4 knobs + 4 buttons like the [standard effects mapping](https://mixxx.org/wiki/doku.php/standard_effects_mapping) for Effect Unit 1 then use the 2x2 grid for deck routing buttons of Effect Unit 1 and the Filter button to switch between Effect Unit 1 and the QuickEffect for each deck. But I don't know how you could combine this with loading effect chain presets to the QuickEffects in 2.4.

ywwg commented:
OK how about I keep it simple:

* knob controls quickeffect only
* fx buttons control enable/disable for the fx units
* filter-enable button on/off: use fx units

Be-ing commented:
> fx buttons control enable/disable for the fx units

Do you mean the enable switch for the whole chain?

> filter-enable button on/off: use fx units

How?

ywwg commented:
> Do you mean the enable switch for the whole chain?

yes that was my thought



> How?

we maintain internal state for the fx units and the fx-enable state of each channel.  When either is enabled or disabled, we go through the channels and set them enabled where both are turned on:

```
for (channel in channels) {
  for (fxUnit in units) {
    engine.setvalue(channel, effects unit, (channel.fx_enabled && fxUnit.enabled));  
  }
}
```

Let's say channels 1 and 2 both have fx enabled.  And FX units 2 and 4 are lit.  Both channels 1 and 2 will use fx units 2 and 4.  if I enable fx on channel 3, it will also use fx units 2 and 4.  if I disable fx unit 4, channels 1, 2, and 3 will be using only fx unit 2.  it's just a simple ANDing of what's lit.

ywwg committed:
Traktor S3: batch light updates for less HID spamming

The wheels and VU meters cause a lot of packets to be sent -- instead,
batch up the values and send them every gui tick.

Also simplify FX routing by only using super knob for the filter.

ywwg committed:
Merge remote-tracking branch 'be/hid_polling' into traktor-s3-2.3

Be-ing commented:
I don't understand how it is useful to toggle effect units on and off without any controls besides that binary toggle.

ywwg commented:
I don't know what you'd suggest -- this is an entry-level controller, that's all I've got to work with.  I don't think "leave it unmapped" is better than "do the best we can with the controls we have"

Holzhaus commented:
I think the issue is that mapping changes are tedious to document on the wiki, because we need to keep documentation for older Mixxx versions too. I'm very much in favor of moving the mapping documentation for built-in mappings to the manual. That way we get versioning for free and only need to document the current behavior. I'll have a look how to accomplish that.

Be-ing commented:
I already made a [suggestion](https://github.com/mixxxdj/mixxx/pull/3031#issuecomment-678802995) above. What do you think of that?

Be-ing commented:
Alternatively you could map the 2x2 grid to switch which effect unit the 4 knobs + 4 buttons control, but that way there would be no access to the routing switches.

ywwg commented:
> 4 knobs + 4 buttons


this seems like a much more unintuitive solution because the knobs and buttons are used for the exact opposite purpose than is indicated by the markings on the controllers.  This would make the buttons marked "fx" select actually select decks, and make the fx-enable buttons actually enable fx.  We have to balance what the controller designer intended with how mixxx works.  In this case, trying to jam the more complex effect model into this controller results in a much more unfriendly experience.

We could preserve some of your idea with some push-and-hold logic:

Say:  by default, knobs control quickeffect.  If I press-and-hold fx1, fx1 blinks and the fxenable buttons glow with the color of fx1.  Now the 4 knobs and buttons operate in standard effects adjustment mode.  Pressing fx1 again disables that mode and returns the knobs to quickeffect.  Furthermore, press-and-hold fx-enable plus tapping an fx select button could allow the user to select any set of filters per deck.

It's a lot of pressing and holding, but the color changes would help guide the user, and even if they don't read any of the wiki, the quickeffect knobs will still work as expected.

ywwg commented:
>  Pressing fx1 again

oh!  no need to press-and-hold.  just tap any fx select and the knobs are in standard effects mapping mode.  if they want quickeffect, they need to tap filter.  So the fxselect buttons are mode selectors for the fx. 

as for routing, push-and-hold on+fx# works fine.   press-and-hold would also light up which effects are enabled.  I think this can work!

ywwg commented:
> Alternatively you could map the 2x2 grid to switch which effect unit the 4 knobs + 4 buttons control, but that way there would be no access to the routing switches.

I think I just suggested what you already suggested here, tho I have a way to do routing

ywwg committed:
Traktor S3: Start rewriting fx mapping

Be-ing commented:
> as for routing, push-and-hold on+fx# works fine. press-and-hold would also light up which effects are enabled. I think this can work!

This could work, but it will be complicated both to implement and use. It would also preclude implementing effect focusing which in the standard effects mapping relies on press-and-holding the button next to the dry/wet knob. I think it would be more straightforward to use the Mixer FX knobs + enable buttons to control a single effect unit and use the 2x2 grid for the routing switches.

> trying to jam the more complex effect model into this controller results in a much more unfriendly experience

I agree, trying to cram controls for the regular effects units into this controller is going to be odd no matter how you do it. The controller is explicitly not designed for this (probably to create more of a difference between the S3 and the much more expensive S4 Mk3).

ywwg committed:
Traktor S3: write a controller unit test!

Very initial commit

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: refactor testing, js testing works now

ywwg commented:
moving back to draft because I'm completely reworking the effects setup

Be-ing commented on src/controllers/controllerengine.cpp:
```diff
@@ -1096,6 +1095,23 @@ bool ControllerEngine::evaluateWithReturn(const QFileInfo& scriptFile, QScriptVa
+   Output:  false if the script file has errors or doesn't exist, and the result
+            of the script if outValue is not nullptr.
+   -------- ------------------------------------------------------ */
+bool ControllerEngine::evaluateWithReturn(const QString& scriptCode,
```
These changes to ControllerEngine are likely going to have to be redone from scratch for master.

Be-ing replied:
I'm not comfortable rushing this new feature into a beta branch, especially considering that it depends on already outdated code. I don't like that a C++ file has to be written for every specific controller mapping with this approach. If it helps you write this mapping, that's fine, but I think we should remove all C++ changes before merging this. I don't really want to work on this new feature while we have a major refactoring of the same part of the code in progress. If you want to discuss how to fit testing into the new controller system, let's discuss that on Zulip before rushing a feature into a beta branch.

ywwg replied:
they're minor, so ok

ywwg replied:
it looks like evaluateCodeString is already the thing I want -- we could commit just the js to 2.3 and commit the js plus tests to master.  or, I could target this to master and then backport just the js to 2.3

Be-ing replied:
I'm not convinced this mixing of C++ and JS is a good approach to testing scripts. I also don't think we should be investing any more effort in the legacy mapping system right now; let's work to build a better system instead. If we add a framework for testing controller scripts, IMO it should be done with self-contained JS files. We can't really do that well right with the legacy mapping system having all mapping files in one big, messy folder. Let's look at common practices for JS testing frameworks and carefully consider the design before building our own.

Be-ing replied:
Let's discuss this [on Zulip](https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping/topic/testing.20controller.20scripts).

ywwg replied:
When a new controller system exists, I'll use it.  For now, the only way I was able to rewrite the effects system was to write these tests.  The amount of change to the legacy infrastructure is quite minimal.

ywwg replied:
For now, what do you think of my proposal to port this to master, and then backport just the controller script to 2.3?

ywwg committed:
Traktor S3: wip -- more effects reworking

ywwg committed:
Traktor S3: add a test for enable/disable effect units per channel

ywwg committed:
TraktorS3: more effects mapping and testing

ywwg committed:
Traktor S3: remove commented cruft

JoergAtGithub commented:
I expect that this mapping is also affected by: https://bugs.launchpad.net/bugs/1893849

ywwg committed:
Traktor S3: start testing lights

ywwg committed:
Traktor S3: more tests

ywwg committed:
Traktor S3: tests pass!

ywwg committed:
Traktor S3: Effects mapping mostly done

TODO:
* shift + knobs in focus mode to select effects

ywwg committed:
Traktor S3: debug cleanup

ywwg committed:
revert unintentional changes

ywwg committed:
another unintentional commit

ywwg commented:
https://youtu.be/8hkrahP9yiY

ywwg commented:
https://youtu.be/Sb13mw06zUM

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: Sync with master version

ywwg committed:
Traktor S3: revert test for 2.3

ywwg committed:
Traktor S3: revert formatting changes

ywwg commented:
Still learning more about the right way to write javascript, so I did another refactoring pass. This is identical to the master version, which is tested.

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: make sure to bind callbacks

ywwg committed:
Traktor S3: Hook up jog button

Press and hold to use wheels to quickly scroll through track

ywwg committed:
Merge branch 'hid-poll-fix' into traktor-s3-2.3

ywwg committed:
Traktor S3: Much better jog/scratch wheel behavior

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

Holzhaus commented on res/controllers/Traktor Kontrol S3.hid.xml:
```diff
@@ -0,0 +1,19 @@
+    <name>Traktor Kontrol S3</name>
+    <author>Owen Williams</author>
+    <description>HID Mapping for Traktor Kontrol S3</description>
+    <wiki>https://www.mixxx.org/wiki/doku.php/native_instruments_traktor_s3</wiki>
```
Oldschool wiki link?

ywwg replied:
fixed

Holzhaus reviewed (commented):
Starting with Mixxx 2.3, all controller mappings are now supposed to be documented in the Manual. This is necessary so that other users know how to use this mapping. We also need it to review it properly and check if the mapping makes sense and if the code matches the documentation.

Please fork the mixxxdj/manual repository, check out the manual-2.3.x branch and add documentation to the source/hardware/controllers directory. Open the pull request and add the link to the PR description.

Here's an example how such documentation could look like: https://manual.mixxx.org/2.3/en/hardware/controllers/roland_dj_505.html (If you don't have schematic images and don't want to draw them yourself, just use the same numbering as in the controller's manual and document the buttons like that).

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: Fix relative slider mode out of bounds issues

ywwg commented:
I fixed the rate slider issue so this is good to go

ywwg commented:
(fixing the wiki link now)

ywwg committed:
Traktor S3: Update wiki link

JoergAtGithub commented:
The location for 2.3 mapping documentation is here: https://www.mixxx.org/manual/2.3/en/hardware/manuals.html not in the Wiki.

ywwg commented:
ah ok

ywwg commented:
added

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

Holzhaus commented on res/controllers/Traktor Kontrol S3.hid.xml:
```diff
@@ -0,0 +1,19 @@
+    <name>Traktor Kontrol S3</name>
+    <author>Owen Williams</author>
+    <description>HID Mapping for Traktor Kontrol S3</description>
+    <wiki>https://github.com/mixxxdj/mixxx/wiki/Native-Instruments-Traktor-Kontrol-S3</wiki>
```
Please use the `<manual>` tag instead.

Holzhaus commented on res/controllers/Traktor Kontrol S3.hid.xml:
```diff
@@ -0,0 +1,19 @@
+    <author>Owen Williams</author>
+    <description>HID Mapping for Traktor Kontrol S3</description>
+    <wiki>https://github.com/mixxxdj/mixxx/wiki/Native-Instruments-Traktor-Kontrol-S3</wiki>
+    <forums></forums>
```
```suggestion
    <forums>https://mixxx.discourse.group/t/native-instruments-traktor-s3-mapping/18502/4</forums>
```

Holzhaus commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2234 @@
+/* Author: Owen Williams                                                         */
+/* https://www.mixxx.org/wiki/doku.php/native_instruments_traktor_kontrol_s3     */
+/*                                                                               */
+/* For linter:                                                                   */
+/* global HIDController, HIDDebug, HIDPacket, controller                         */
```
Please add this to the eslint config if necessary.

Holzhaus commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2234 @@
+///////////////////////////////////////////////////////////////////////////////////
+/*                                                                               */
```
Wouldn't it make more sense to use `//` instead?

Holzhaus commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2234 @@
+
+TraktorS3.Deck.prototype.numberButtonHandler = function(field) {
+    var padNumber = parseInt(field.name[field.name.length - 1]);
+    var action = "";
+
+    // Hotcues mode
+    if (this.padModeState === 0) {
+        if (this.shiftPressed) {
+            action = "_clear";
+        } else {
+            action = "_activate";
+        }
```
How about this? 

```suggestion

    // Hotcues mode
    if (this.padModeState === 0) {
        var action = this.shiftPressed ? "_clear" : "_activate";
```

Holzhaus reviewed (changes requested):
Mostly looks good, thank you. I obviously can't test because I don't have the hardware.

aartigao commented:
First of all, amazing work @ywwg 🙌🏽 I've been testing for an hour or so and the mappings are impressive 🚀 

There are though some differences that I see comparing with **Traktor Pro 3**s default mappings. For example, `JOG` button in **Traktor Pro 3** disables the touch for scratch and the rest behaves the same. This is a minor thing for me, but I wanted to point out because I'm new to **Mixxx** and I don't know if the purpose of the mappings is to behave as much as possible as the original software or otherwise you just fit the hardware into **Mixxx**'s features.

That being said, there's one thing that I noticed that I think is not working as I expected (or maybe I didn't find the correct combination): 

- The 🎧 / `MIX` knob and the 🎧 `CUE` buttons doesn't seem to work together. I would expect to hear in headphones any Deck with 🎧 `CUE` activated, but I don't (I'm forced to have the channel fader active too and same happens with global fader). I suspect what is missing is the implementation for routing between `CUE` and `MASTER` channels 🤔 

Can you confirm if this is the expected behavior?

In the following days I'll be testing other parts like `HOTCUES` / `SAMPLES`. I'll keep you posted on my progress 😄

Be-ing commented:
> I would expect to hear in headphones any Deck with 🎧 CUE activated

That's only with the headphone/main mix knob turned towards headphones. Is that what you mean?

It's possible the audio interface mixes the headphone and main signals in hardware so mapping the knob to do that in Mixxx is redundant.

Holzhaus commented:
> It's possible the audio interface mixes the headphone and main signals in hardware so mapping the knob to do that in Mixxx is redundant. If that's the case, the knob should not be mapped.

If that is the case, that needs to be documented on the manual page: https://deploy-preview-308--mixxx-manual.netlify.app/index.html

ywwg commented:
headphone mix and cue buttons should work.  They might have soft takeover active which would make it seem like they aren't mapped right.

ywwg committed:
Update res/controllers/Traktor Kontrol S3.hid.xml

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

Be-ing commented:
@ywwg can you test if the headphone/main mix knob works if you comment out the mapping? If so, then don't map it.

ywwg committed:
Update res/controllers/Traktor-Kontrol-S3-hid-scripts.js

Co-authored-by: Jan Holthuis <holthuis.jan@googlemail.com>

ywwg committed:
Traktor S3: address notes

ywwg committed:
Merge branch 'traktor-s3-2.3' of github.com:ywwg/mixxx into traktor-s3-2.3

aartigao commented:
> headphone mix and cue buttons should work. They might have soft takeover active which would make it seem like they aren't mapped right.

Checking the source code those lines are commented. But anyway I tried uncommenting and retry without luck 😞 

> @ywwg can you test if the headphone/main mix knob works if you comment out the mapping? If so, then don't map it.

I tried that too, but still no luck... 🤷🏽

Be-ing commented:
> I tried that too, but still no luck... 🤷🏽

Can you clarify? Did the headphone/mix knob on the controller not work at all when you did that? Or did it behave the same way as before? Can you post a screenshot of your Sound Hardware preferences?

aartigao commented:
> Can you clarify? Did the headphone/mix knob on the controller not work at all when you did that? Or did it behave the same way as before?

It behaved the same way as before.

> Can you post a screenshot of your Sound Hardware preferences?

![Screenshot from 2020-12-07 18-05-22](https://user-images.githubusercontent.com/6449814/101381315-e6a97280-38b6-11eb-87e4-ae0c5918887a.png)

Be-ing commented:
Oh, I assumed you had your audio I/O set up correctly. You need to configure the S3's audio interface for the Master and Headphones outputs. https://manual.mixxx.org/2.2/en/chapters/preferences.html#preferences-sound-hardware

Holzhaus commented:
The sound configuration should documented on the manual page. @ywwg can you add a little table what channels should be assigned to what output/input?

ywwg commented:
added: https://deploy-preview-308--mixxx-manual.netlify.app/hardware/controllers/native_instruments_traktor_kontrol_s3.html

Be-ing commented:
> There are though some differences that I see comparing with Traktor Pro 3s default mappings. For example, JOG button in Traktor Pro 3 disables the touch for scratch and the rest behaves the same. This is a minor thing for me, but I wanted to point out because I'm new to Mixxx and I don't know if the purpose of the mappings is to behave as much as possible as the original software or otherwise you just fit the hardware into Mixxx's features.

That's odd. Please change that so the jog button works like other controllers: nudging when off, scratching when on. Fast scrolling with jog wheels is conventionally done with shift + wheel turn, so that feature doesn't need to be removed.

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+
+    for (ch in this.Channels) {
+        var chanob = this.Channels[ch];
+        engine.connectControl(ch, "playposition",
```
```suggestion
        engine.makeConnection(ch, "playposition",
```

`engine.connectControl` is deprecated

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+    var data = [Array(), Array()];
+
+
+    for (var i = 0; i < data.length; i++) {
+        var tokens = dataStrings[i].split(/\s+/);
+        data[i].length = tokens.length;
+        for (var j = 0; j < tokens.length; j++) {
+            var byteStr = tokens[j];
+            if (byteStr.length === 0) {
+                continue;
+            }
+            data[i][j] = parseInt(byteStr, 16);
+        }
+        controller.send(data[i], data[i].length, 0x80 + i);
+    }
```
This is quite a complex way to fill an array with 0s.

ywwg replied:
You make an excellent point, I'll fix this

Holzhaus replied:
ping?

ywwg replied:
fixed

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+TraktorS3.debugLights = function() {
+    // Call this if you want to just send raw packets to the controller (good for figuring out what
+    // bytes do what).
+    var dataStrings = [
+        "00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "22 22 22 22  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "22 22 22 22  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "FF FF FF FF  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00 00 00 00 ",
+        "00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00 00 00 00  00 00 00 00  00 00 00 00  00 00 00",
+        "20 08 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00 " +
+        "00"
+    ];
```
Why not use a normal JS Array?

ywwg replied:
I do it like this because I am cutting and pasting the hex values from my USB sniffing program.  This way I can just paste in the values and run the code to see that the lights work the same was as in traktor.

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+    if (TraktorS3.DebugMode) {
+        TraktorS3.debugLights();
+    } else {
+        this.kontrol.lightDeck("[Channel3]", false);
+        this.kontrol.lightDeck("[Channel4]", false);
+        this.kontrol.lightDeck("[Channel1]", false);
+        this.kontrol.lightDeck("[Channel2]", true);
```
Is there any particular reason for this unusual order?

ywwg replied:
yes!  comment added

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+        return false;
+    }
+    this.positionUpdated = false;
+    var rotations = this.curPosition * (1 / 1.8);  // 1/1.8 is rotations per second
```
```suggestion
    var rotations = this.curPosition * (1 / 1.8);  // 1/1.8 is rotations per second (33 1/3 RPM)
```

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+    var angle = rotations - Math.floor(rotations);
+    // The wheel has 8 segments
+    var wheelAngle = 8.0 * angle;
+    var baseLedValue = this.channelBaseColor();
```
I'm going to have to try making my S4 Mk3 spin rainbows around the wheels. :laughing:

Be-ing replied:
Then make a video with it while isolating a rainbow LED hoop :rainbow:

Be-ing replied:
If you want to add an easter egg option for this, that would be cool. Of course you don't have to.

ywwg replied:
that would be very fun, but unfortunately my code assumes 1 color for all the wheel segments... I'll see if there's an easy way to change that, but maybe that's something to save for a subsequent commit :)

Be-ing replied:
You could have one spot unlit to indicate the track position.

ywwg replied:
ooo, inverting colors does sound like a good idea.  I'll try it

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+    // The wheel has 8 segments
+    var wheelAngle = 8.0 * angle;
+    var baseLedValue = this.channelBaseColor();
+    // Reduce the dimming distance at the end of track.
```
That's an interesting way to show that information.

ywwg replied:
CDJs flash at end of track, and I found that I needed to tweak the wheel segments to make it obvious that the wheel was flashing.  So this is my own design based on what looked best / most easy to quickly understand

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+    // How many segments away from the actual angle should we light?
+    // (in both directions, so "2" will light up to four segments)
+    if (this.trackDurationSec === 0) {
+        var samples = engine.getValue(this.group, "track_samples");
```
```suggestion
        var samples = engine.getValue(this.group, "track_loaded");
```
There's a ControlObject specifically for this purpose now.

ywwg replied:
cool, thanks

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+};
+
+TraktorS3.Deck.prototype.pitchSliderHandler = function(field) {
+    // Adapt midi value to rate control range.
```
```suggestion
    // Adapt HID value to rate control range.
```

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+        "[Channel3]": false,
+        "[Channel4]": false
+    };
+    this.selectPressed = {
+        0: false,
+        1: false,
+        2: false,
+        3: false,
+        4: false
+    };
```
Why not use an Array instead of an Object for this?

ywwg replied:
probably leftover from some old code.  fixed

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -0,0 +1,2227 @@
+    if (this.shiftPressed) {
+        engine.setValue(this.activeChannel, "start_stop", field.value);
+    } else if (field.value === 1) {
+        script.toggleControl(this.activeChannel, "play");
```
Does this need to cancel the wheel inertia timer and/or stop scratching? I had to do that for the Kontrol S2 Mk2 mapping or it got stuck in a weird state where the deck wouldn't play after doing a backspin and quickly pressing the play button. I don't know if that's needed with the way you're directly manipulating the `scratch2` Control.

Be-ing replied:
https://github.com/mixxxdj/mixxx/blob/main/res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js#L629

ywwg replied:
I'll check this out

Be-ing commented:
I haven't read through every detail of all 2200 lines of code but it looks pretty good. I noted a handful of comments above. My biggest concerns before merging this are:
  * [ ] clarify effects documentation
  * [ ] add pad grid documentation
  * [ ] fixing jog/scratch toggle button

Be-ing commented:
Thank you for your patience with this review. I've been focused on 2.3 release blockers.

Be-ing commented:
This seems like a nice controller to recommend for users who want quality hardware without spending $1000.

Be-ing commented:
Out of curiosity, I tested this with my S4 Mk3. LEDs light up in random colors. The knobs and faders of the mixer section move knobs and faders, but they're not the right ones.

Be-ing commented:
Does the S3 have a way to adjust the jog wheel tension? That might only be on the S4 Mk3.

ywwg commented:
> Does the S3 have a way to adjust the jog wheel tension? That might only be on the S4 Mk3.

no there isn't

ywwg commented:
> This seems like a nice controller to recommend for users who want quality hardware without spending $1000.

the build quality isn't great, and effects manipulation is awkward, but the basic controls are quite nice.

ywwg committed:
Merge remote-tracking branch 'mixxx/2.3' into traktor-s3-2.3

ywwg committed:
Update res/controllers/Traktor-Kontrol-S3-hid-scripts.js

Co-authored-by: Be <be.0@gmx.com>

Be-ing commented:
Please write a useful commit message when applying suggestions directly from GitHub reviews.

ywwg committed:
Traktor S3: address notes

ywwg committed:
Merge remote-tracking branch 'origin/traktor-s3-2.3' into traktor-s3-2.3

ywwg committed:
Apply suggestions from code review

Co-authored-by: Be <be.0@gmx.com>

ywwg committed:
Merge remote-tracking branch 'origin/traktor-s3-2.3' into traktor-s3-2.3

ywwg commented:
> I haven't read through every detail of all 2200 lines of code but it looks pretty good

thanks for taking a look, I know it's long.  We don't really have a way to commit partial controller mappings unless it's like, "just knobs and sliders" "add lights" "add effects"

Be-ing commented:
> We don't really have a way to commit partial controller mappings unless it's like, "just knobs and sliders" "add lights" "add effects"

That's fine. I wouldn't want partial controller mappings merged. We already have the forums for that.

Holzhaus commented:
Yes, it still possible to split stuff into separate commits for easier review.

Holzhaus commented:
What's the status of this? Do you consider this ready to merge? In any case there are dome open issues on https://github.com/mixxxdj/manual/pull/308 that need to be resolved before merging.

Be-ing commented:
> What's the status of this?

https://github.com/mixxxdj/mixxx/pull/3031#issuecomment-740391403

ywwg commented:
Yeah I have to address those comments.  I am on vacation starting friday, so I should be able to get it done.  Are we gearing up for a classic Mixxxmas release??? :)

Be-ing commented:
That seems to be how it's going.

ywwg commented:
>  nudging when off, scratching when on.

I'm remapping so the shift button does a seek.  But I'm not sure that's the best use of the jog button.  This deck has touch detection, so you touch the edge if you want to nudge, and touch the top platter to scratch. This matches how CDJs work.  Therefore a separate button doesn't seem necessary.

ywwg commented:
I'll do it that way anyway for now but I feel like it could be used more effectively some other way.

Be-ing commented:
IIUC, the issue that people seek to avoid with such a switch is accidentally stopping the track by touching the top of the jog wheel.

ywwg commented:
> IIUC, the issue that people seek to avoid with such a switch is accidentally stopping the track by touching the top of the jog wheel.

Are you saying the mapping should not respond to the platter touch?  That's counter to the way this controller is meant to be used and the way most jog wheels behave.  There's a lot of buttons a DJ could accidentally hit that might cause an undesired result (pause, eq kill, hotcue, loop...), so it's not Mixxx's job to prevent accidental presses.  (The first time I used a CDJ the platter touch was confusing, but I got used to it very quickly). We could use a controller preference, I guess, but I'd prefer to encourage people to learn the mapping in a way that will serve them the best if they move to other controllers or professional gear.

aartigao commented:
Just a question about Jog, have you tried it in Traktor Pro? I can't tell 100% because right now I don't have the S3 here to test but IIRC in TP works as described by @Be-ing (at least by default)

Be-ing commented:
AFAIK just about every controller with a jog wheel works that way...

Be-ing commented:
Ping @ywwg, will you have time to finish this soon? It would be nice to have it merged for 2.3.0. If not, that's no big deal, we can easily make a point release when this is merged.

ywwg commented:
Yes I'll try to address the notes this weekend

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: Create options for different sampler playback modes

ywwg commented:
I have to doublecheck that I fixed the jog button, but the manual items are addressed

ywwg commented:
I still have to update the jog button behavior

ywwg committed:
Traktor S3: Make JOG button work like it does for other controllers

quendera commented:
Hey there @ywwg , I have tested this PR and have a few comments to make about user experience. I am new to mixxx so apologies if I am too vague. 
Overall usability is great, thank you for the amazing work. I have a few comments:

1. The automatic build doesn't run on Mint 20 kernel 5.4.0-65-generic. When compiled from source worked fine.
2. The effects controls are very weird. couldn't figure out from the docs page how to actually use fx with the S3 
3. fx rack is also buggy at times e.g. leftmost track toggle is always on, some fx leds  (couldn't find a pattern) never turn off.
4. 4 deck mode has same colored LEDs for decks AB and CD.
5. view button changes from tree to folder view. that is problematic in 4 deck mode because you lose the ability to view, without resorting to mouse, playlists/tracklist. I think view should allow you to "view" the tracks (like in traktor), and maybe the list button (unused) could toggle tree/folder view.
6. It's intuitive to me that touching jog on top should pause track - vinyl mode (like cdj or traktor). couldn't figure out how to get this working.

I'll add more comments later today as I test it a bit further.
I will also try to understand how you coded the fx rack and see if I can make a PR with some suggestions to the mapping or the docs.

Thanks again for the amazing work.
Let me know how I can best help you.

Be-ing commented:
Thank you for testing @quendera. It is always better to have people who have the hardware give feedback. The rest of us can only infer what it is like to use from the documentation. If you want to make changes, make a PR for @ywwg's fork and leave a comment here linking to your PR.

ywwg committed:
Traktor S3: simplify shutdown

ywwg commented:
thanks for your feedback.  When you say the LEDs are the same for AB as CD, are you saying when you push the deck select button that nothing changes?

I will look at the effects issues.  Updated manual is here: https://github.com/mixxxdj/manual/pull/308

I don't have traktor, what does the View button do?

>     6\. It's intuitive to me that touching jog on top should pause track - vinyl mode (like cdj or traktor). couldn't figure out how to get this working.

this is what I was saying before -- can we restore this as the default?

Be-ing commented:
> this is what I was saying before -- can we restore this as the default?

I don't care what the default is. The issue is the button to change this only takes effect while held down. This should be a button that toggles between modes.

ywwg committed:
Traktor S3: Small updates:

* Fix library view switching
* Fix shift+wheel not releasing right away
* add shift+right encoder for moving loops

quendera commented:
>  When you say the LEDs are the same for AB as CD, are you saying when you push the deck select button that nothing changes?

I meant that A and B are both orange while C and D are blue. This is minor, but it did confuse me because the center FX toggles are also in this color. Changing FX button 3 to a color different than blue would help clear it for me.


> I will look at the effects issues. Updated manual is here: mixxxdj/manual#308

After testing further and watching the youtube videos from @ywwg in this thread, I understood how the fx system currently works. 

It's extremely convoluted as the controller fits best with a single effect for all knobs. That said, it adds a lot of functionality to the controller so I thank you very much for this.
If you are comfortable with that, adding your demo video to the manual could help a lot of people.

I found out my inconsistency with the LEDs in fx mode - leds corresponding to effects in the rack only light up after you exit and re-enter the effect. When no effects are attributed to the rack, the leftmost led still stays on.
 
The weird thing as of now is that, once you enter FX rack mode (by pressing one of the number buttons on fx select), the knobs no longer control the tracks' effect but the racks'. Thus, the rightmost knob (in track D) controls the masterknob for that given fx rack - Is this the case?

Later next week I can try to help you by editing the manual too.

> I don't have traktor, what does the View button do?

View toggles big library mode. After testing, the currently implemented feature is indeed very useful and it could be done using the currently unused button to its left.
![image](https://user-images.githubusercontent.com/21177133/107130698-20532580-68c8-11eb-9455-1e01beabed3a.png)

quendera commented:
- The status of encoders, knobs and faders is not queried on opening the software. For example, on mixxx all track faders are fully up even if controller has the faders down, gain is not the same as my gain knobs.
Is this the intended behaviour?
Since I never used mixxx with another controller, unsure whether this is the place to comment on or if it's a general feature/limitation from mixxx.

- I don't understand how the jog wheels work. In other software/hardware I've used (serato, traktor or cdj) touching the top of the controller pauses the track and, as soon as released, resumes playing it - basically, I would expect it to do the same as if you hold your mouse on the mixxx waveforms. 
At the same time, pressing jog button does stop the track. I find this very confusing as I don't see what's the use of jog mode if I can't have the track playing while it's active.

Be-ing commented:
> The status of encoders, knobs and faders is not queried on opening the software.

I think this requires a feature that won't be in 2.3: https://github.com/mixxxdj/mixxx/pull/3074#discussion_r571473582

ywwg commented:
> Changing FX button 3 to a color different than blue would help clear it for me.

Maybe I can make the FX button purple instead.  I'll try to make them more different. Unfortunately there are only 16 colors to work with.

> The issue is the button to change this only takes effect while held down. This should be a button that toggles between modes.

yes, that's what the button does now.

> View toggles big library mode. After testing, the currently implemented feature is indeed very useful and it could be done using the currently unused button to its left.

I believe in traktor that the View button toggles Big Library mode.  I will do that here, and make the button to the side move the focus around.  

I think I will make the Star button toggle Played status.

ywwg committed:
Traktor S3: remap "view" to maximize_library, and playlist button to move focus

ywwg commented:
> I think I will make the Star button toggle Played status.

there is no control object for this, so I can't do it

ywwg commented:
>     * touching the top of the controller pauses the track and, as soon as released, resumes playing it

I've restored this as the default behavior, try it out again!

>  found out my inconsistency with the LEDs in fx mode - leds corresponding to effects in the rack only light up after you exit and re-enter the effect. When no effects are attributed to the rack, the leftmost led still stays on.

I'll see if I can reproduce this.

> 
> The weird thing as of now is that, once you enter FX rack mode (by pressing one of the number buttons on fx select), the knobs no longer control the tracks' effect but the racks'. Thus, the rightmost knob (in track D) controls the masterknob for that given fx rack - Is this the case?

That is the intended  behavior.  Once you push an fx select button, the knobs and buttons now control the details of that effect specifically.  I realize it's a lot of overloaded behavior... I was thinking of having a Simple mode that could be the default that would just allow turning on/off effects for tracks, and the fx selects would apply to all tracks (which is how it works in traktor, according to the manual).

Be-ing commented on res/controllers/Traktor-Kontrol-S3-hid-scripts.js:
```diff
@@ -37,6 +37,11 @@ TraktorS3.PitchSliderRelativeMode = true;
+// By default, touching the jog wheel does nothing special.  When this option is true,
+// touching the job wheel enables scratch mode. Pressing the jog button still also
+// enables scratching.
+TraktorS3.WheelTouchScratching = false;
```
You've still misunderstood. This should not be a static set-and-forget option. It should be a state toggled by the jog button. You can keep this preference option though for the user to set their on default state on startup.

ywwg replied:
oh! got it

ywwg committed:
Merge branch '2.3' into traktor-s3-2.3

ywwg committed:
Traktor S3: Fix so jog button toggles wheel touch mode

ywwg commented:
fixed, and updated manual PR too

Be-ing commented:
Okay, the only major issue remaining IMO is clarifying the effects documentation (and addressing my pending comments on the manual PR).

Be-ing commented:
Thank you @ywwg. @aartigao, @quendera, if you want to propose further changes to this mapping feel free to make follow up pull requests. I suggest discussing them first in the controller mapping stream on [Zulip](https://mixxx.zulipchat.com/#narrow/stream/113295-controller-mapping).
